functionary/prompt_template/base_template.py

```diff
@@ -11,6 +13,393 @@
+        self,
+        gen_state: Dict,
+        new_token_id: int,
+        options: Optional[List],
+        tokenizer: Any,
+    ) -> Dict:
+        """Receives a generation state, updates and returns it. This is only used when
+        grammar sampling is enabled in inference. This functions parses the generated
+        tokens and identifies the stage of generation (pre-function, function, parameter-name,
+        etc.)
+        Args:
+            gen_state (Dict): The current generation state. It contains the following:
+            - stage: one of the following:
+              - pre-function: the generation prior to function name generation
+              - function: when the model is generating a function name
+              - pre-parameter: when the model is generating the part between function name and parameter
+              - parameter-name: when the model is generating a parameter name
+              - parameter-value: when the model is generating a parameter value
+              - no-function-call: when the model is generating content
+            - curr_tokens: all the tokens for the current stage being generated
+            - curr_text: curr_tokens but in string text form
+            - func_name: the function name, if any
+            - param_names: the parameters names, if any
+            new_token_id (int): The token id of the newly sampled token
+            options (List): All available function/param names depending on the stage of gen_state
+            tokenizer (Any): The tokenizer class passed in from Transformers or vLLM
+        Returns:
+            dict: The updated gen_state
+        """
+        # Update curr_tokens and curr_text
+        gen_state["curr_tokens"].append(new_token_id)
+        gen_state["curr_text"] = tokenizer.decode(gen_state["curr_tokens"])
+
+        # v1: "assistant:\n{content}\n{self.start_function}{function}:\n{arguments}\n"
+        # v2: "{func_name}\n<content|>{param_names}\n<|from|> assistant\n<|recipient|>"
+        if gen_state["stage"] == "pre-function":
+            # Check if the new state is in "function" stage
+            if gen_state["curr_text"].endswith(self.get_start_of_function_call_token()):
+                gen_state = {
+                    "stage": "function",
+                    "curr_tokens": [],
+                    "curr_text": "",
+                    "func_name": "",
+                    "param_names": [],
+                }
+                gen_state["stage"] = "function"
+        elif gen_state["stage"] == "function":
+            # Remove all unnecessary suffixes by checking whether stop token is in curr_text
+            if (
+                self.get_stop_token_for_function_parameter(stage="function")
+                in gen_state["curr_text"]
+            ):
+                curr_text = gen_state["curr_text"].rstrip()
+                while True:
+                    if any([curr_text == option for option in options]):
```

Comment:
This while loop will not be forever because the `grammar_sample` function already forces the function name to be built towards one of the provided function names including "all". This part just loops from the back to remove unnecessary suffixes until we have the complete function name to put into gen_state["func_name"]